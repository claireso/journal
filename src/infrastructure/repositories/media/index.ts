export { default as MediaRepositoryImpl } from './MediaRepositoryImpl'
