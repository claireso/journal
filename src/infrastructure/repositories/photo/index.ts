export { default as PhotoRepositoryImpl } from './PhotoRepositoryImpl'
