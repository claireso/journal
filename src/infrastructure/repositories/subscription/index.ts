export { default as SubscriptionRepositoryImpl } from './SubscriptionRepositoryImpl'
