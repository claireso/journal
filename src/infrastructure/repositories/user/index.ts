export { default as UserRepositoryImpl } from './UserRepositoryImpl'
