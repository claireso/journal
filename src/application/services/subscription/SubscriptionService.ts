import { NotFoundError } from '@domain/errors/errors'
import { Pager, Subscriptions } from '@domain/entities'
import { SubscriptionRepository } from '@domain/repositories'

export default class SubscriptionService {
  private repository: SubscriptionRepository
  private logger: unknown

  constructor(repository: SubscriptionRepository, logger: unknown) {
    this.repository = repository
    this.logger = logger
  }

  async create(data: object) {
    return this.repository.create(data)
  }

  async getById(id: number) {
    const subscription = await this.repository.getById(id)

    if (subscription === null) {
      throw new NotFoundError(`Subscription not found`, { cause: { subscriptionId: id } })
    }

    return subscription
  }

  async getAll() {
    return this.repository.getAll()
  }

  async delete(id: number) {
    const subscription = await this.getById(id)

    return this.repository.delete(subscription)
  }

  async getPaginatedSubscriptions(page: number) {
    const pageSize = 10

    const count = await this.repository.countSubscriptions()
    const totalPages = Math.ceil(count / pageSize)

    if (page < 1 || (totalPages && page > totalPages)) {
      throw new NotFoundError('Page subscription not found', { cause: { page } })
    }

    const offset = (page - 1) * pageSize

    const subscriptions = await this.repository.getSubscriptions(offset, pageSize)

    const pager: Partial<Pager> = {
      count,
      totalPages,
      offset,
      limit: pageSize
    }

    if (page > 1) {
      pager.prev = page - 1
      pager.first = 1
    }

    if (page < totalPages) {
      pager.next = page + 1
      pager.last = totalPages
    }

    return {
      items: subscriptions,
      pager
    } as Subscriptions
  }

  async getSubscriptionByEndpoint(endpoint: string) {
    const registration = await this.repository.getSubscriptionByEndpoint(endpoint)

    if (!registration) {
      throw new NotFoundError('Subscription not found', { cause: { endpoint } })
    }

    return registration
  }
}
