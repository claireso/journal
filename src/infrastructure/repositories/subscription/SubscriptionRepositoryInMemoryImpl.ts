import { Subscription } from '@domain/entities'
import { SubscriptionRepository } from '@domain/repositories'
import { SubscriptionInsertDto } from '@dto'

export default class SubscriptionRepositoryInMemoryImpl implements SubscriptionRepository {
  private subscriptions: Subscription[]

  constructor() {
    this.subscriptions = [
      {
        id: 1,
        created_at: new Date('2024-10-05T17:12:40.400Z'),
        updated_at: new Date('2024-10-05T17:12:40.400Z'),
        subscription: {
          keys: {
            p256dh: '',
            auth: ''
          },
          endpoint: 'https://fcm.googleapis.com/fcm/send/1',
          expirationTime: 1234
        }
      },
      {
        id: 2,
        created_at: new Date('2024-10-05T17:12:40.400Z'),
        updated_at: new Date('2024-10-05T17:12:40.400Z'),
        subscription: {
          keys: {
            p256dh: '',
            auth: ''
          },
          endpoint: 'https://fcm.googleapis.com/fcm/send/2',
          expirationTime: 1234
        }
      },
      {
        id: 3,
        created_at: new Date('2024-10-05T17:12:40.400Z'),
        updated_at: new Date('2024-10-05T17:12:40.400Z'),
        subscription: {
          keys: {
            p256dh: '',
            auth: ''
          },
          endpoint: 'https://fcm.googleapis.com/fcm/send/3',
          expirationTime: 1234
        }
      }
    ]
  }

  async create(data: SubscriptionInsertDto): Promise<Subscription> {
    const lastSubscription = this.subscriptions.at(-1) as Subscription
    const newSubscription = {
      id: lastSubscription.id + 1,
      created_at: new Date('2024-10-05T17:12:40.400Z'),
      updated_at: new Date('2024-10-05T17:12:40.400Z'),
      ...data
    }
    this.subscriptions.push(newSubscription)
    return newSubscription
  }

  async getById(id: number): Promise<Subscription | null> {
    const subscription = this.subscriptions.find((_subscription) => _subscription.id === id)
    return subscription || null
  }

  async getAll(): Promise<Subscription[]> {
    return this.subscriptions
  }

  async delete(subscription: Subscription): Promise<void> {
    this.subscriptions = this.subscriptions.filter((_subscription) => _subscription.id !== subscription.id)
  }

  async getSubscriptions(offset: number, limit: number) {
    const subscriptions = this.subscriptions.slice(offset, limit)
    return subscriptions
  }

  async countSubscriptions(): Promise<number> {
    return this.subscriptions.length
  }

  async getSubscriptionByEndpoint(endpoint: string): Promise<Subscription | null> {
    const subscription = this.subscriptions.find((_subscription) => _subscription.subscription.endpoint === endpoint)
    return subscription ?? null
  }
}
