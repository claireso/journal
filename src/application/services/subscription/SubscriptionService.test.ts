import { SubscriptionRepository } from '@domain/repositories'
import SubscriptionRepositoryInMemoryImpl from '@infrastructure/repositories/subscription/SubscriptionRepositoryInMemoryImpl'
import SubscriptionService from './SubscriptionService'

describe('application/SubscriptionService', () => {
  let subscriptionRepository: SubscriptionRepository
  let subscriptionService: SubscriptionService

  beforeEach(() => {
    subscriptionRepository = new SubscriptionRepositoryInMemoryImpl()
    subscriptionService = new SubscriptionService(subscriptionRepository, console)
  })

  describe('CREATE', () => {
    it('should create a subscription successfully', async () => {
      // arrange
      const data = {
        subscription: {
          keys: {
            p256dh: '123',
            auth: '456'
          },
          endpoint: '789',
          expirationTime: 1234
        }
      }

      // act
      const subscription = await subscriptionService.create(data)

      // assert
      expect(subscription).toMatchSnapshot()
    })
  })

  describe('GET', () => {
    it('should get a subscription by identifier', async () => {
      // act
      const subscription = await subscriptionService.getById(1)
      // assert
      expect(subscription).toMatchSnapshot()
    })
    it('should throw a NotFoundError error when getting a subscription with invalid id', async () => {
      // act
      const promise = subscriptionService.getById(1000)
      // arrange
      await expect(promise).rejects.toThrow('Subscription not found')
    })
    it('should get paginated subscriptions', async () => {
      // act
      const paginatedSubscriptions = await subscriptionService.getPaginatedSubscriptions(1)
      //assert
      expect(paginatedSubscriptions).toMatchSnapshot()
    })
    it('should throw a NotFoundError error when getting paginated subscriptions with invalid page', async () => {
      // act
      const promise = subscriptionService.getPaginatedSubscriptions(1000)
      //assert
      expect(promise).rejects.toThrow('Page subscription not found')
    })
    it('should get all subscriptions', async () => {
      // act
      const subscriptions = await subscriptionService.getAll()
      //assert
      expect(subscriptions).toMatchSnapshot()
    })
    it('should get subscription by endpoint', async () => {
      // arrange
      const endpoint = 'https://fcm.googleapis.com/fcm/send/1'
      // act
      const subscription = await subscriptionService.getSubscriptionByEndpoint(endpoint)
      //assert
      expect(subscription).toMatchSnapshot()
    })
    it('should not get subscription by endpoint', async () => {
      // arrange
      const endpoint = 'https://fcm.googleapis.com/fcm/send/1000'
      // act
      const promise = subscriptionService.getSubscriptionByEndpoint(endpoint)
      //assert
      expect(promise).rejects.toThrow('Subscription not found')
    })
  })

  describe('DELETE', () => {
    it('should delete subscription successfully', async () => {
      // act
      await subscriptionService.delete(1)

      // assert
      expect(subscriptionService.getById(1)).rejects.toThrow('Subscription not found')
    })

    it('should throw a NotFoundError error when deleting subscription with invalid id', async () => {
      // assert
      expect(subscriptionService.delete(1000)).rejects.toThrow('Subscription not found')
    })
  })
})
