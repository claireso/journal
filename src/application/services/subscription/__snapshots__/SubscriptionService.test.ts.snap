// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`application/SubscriptionService CREATE should create a subscription successfully 1`] = `
{
  "created_at": 2024-10-05T17:12:40.400Z,
  "id": 4,
  "subscription": {
    "endpoint": "789",
    "expirationTime": 1234,
    "keys": {
      "auth": "456",
      "p256dh": "123",
    },
  },
  "updated_at": 2024-10-05T17:12:40.400Z,
}
`;

exports[`application/SubscriptionService GET should get a subscription by identifier 1`] = `
{
  "created_at": 2024-10-05T17:12:40.400Z,
  "id": 1,
  "subscription": {
    "endpoint": "",
    "expirationTime": 1234,
    "keys": {
      "auth": "",
      "p256dh": "",
    },
  },
  "updated_at": 2024-10-05T17:12:40.400Z,
}
`;

exports[`application/SubscriptionService GET should get all subscriptions 1`] = `
[
  {
    "created_at": 2024-10-05T17:12:40.400Z,
    "id": 1,
    "subscription": {
      "endpoint": "",
      "expirationTime": 1234,
      "keys": {
        "auth": "",
        "p256dh": "",
      },
    },
    "updated_at": 2024-10-05T17:12:40.400Z,
  },
  {
    "created_at": 2024-10-05T17:12:40.400Z,
    "id": 2,
    "subscription": {
      "endpoint": "",
      "expirationTime": 1234,
      "keys": {
        "auth": "",
        "p256dh": "",
      },
    },
    "updated_at": 2024-10-05T17:12:40.400Z,
  },
  {
    "created_at": 2024-10-05T17:12:40.400Z,
    "id": 3,
    "subscription": {
      "endpoint": "",
      "expirationTime": 1234,
      "keys": {
        "auth": "",
        "p256dh": "",
      },
    },
    "updated_at": 2024-10-05T17:12:40.400Z,
  },
]
`;

exports[`application/SubscriptionService GET should get paginated subscriptions 1`] = `
{
  "items": [
    {
      "created_at": 2024-10-05T17:12:40.400Z,
      "id": 1,
      "subscription": {
        "endpoint": "",
        "expirationTime": 1234,
        "keys": {
          "auth": "",
          "p256dh": "",
        },
      },
      "updated_at": 2024-10-05T17:12:40.400Z,
    },
    {
      "created_at": 2024-10-05T17:12:40.400Z,
      "id": 2,
      "subscription": {
        "endpoint": "",
        "expirationTime": 1234,
        "keys": {
          "auth": "",
          "p256dh": "",
        },
      },
      "updated_at": 2024-10-05T17:12:40.400Z,
    },
    {
      "created_at": 2024-10-05T17:12:40.400Z,
      "id": 3,
      "subscription": {
        "endpoint": "",
        "expirationTime": 1234,
        "keys": {
          "auth": "",
          "p256dh": "",
        },
      },
      "updated_at": 2024-10-05T17:12:40.400Z,
    },
  ],
  "pager": {
    "count": 3,
    "limit": 10,
    "offset": 0,
    "totalPages": 1,
  },
}
`;
